### ISSUE TITLE
Can't access namespaced translations using the filter

### ISSUE THREAD

**Issue by froger-me** [angular-translate/angular-translate#1526]
### Subject of the issue
Can't access namespaced translations using the filter

### Environment
* version of angular-translate : latest following the guide
* version of angular : latest stable
* which browser and its version : chrome latest, latest crosswalk on Android

### Description

When using angular-translate I can filter my strings like so:

`<ANY>{{'TRANSLATION_ID' | translate}}</ANY>`

It works very well with the following flat JSON:

```
 {
  TRANSLATION_ID: 'A value',
};
```
However, I can't figure out how to get it to work using nested translations:

`<ANY>{{'NAMESPACE.TRANSLATION_ID' | translate}}</ANY>`

And the nested JSON:

```
{
    "NAMESPACE": {
         "TRANSLATION_ID": "A value"
    }
};
```
I am successfully loading the translation using `$translateProvider.useStaticFilesLoader({prefix: 'i18n/locale-', suffix: '.json'});`, and I make use of other functions of the provider such as `$translateProvider.addInterpolation('$translateMessageFormatInterpolation');`.

The documentation shows examples of nested translation using the service directly, but not the filter, as far as I've seen.

I found a possibly related issue [here on stackoverflow](http://stackoverflow.com/questions/33537395/angular-translate-using-usestaticfilesloader-to-load-nested-json). First of all, using the code in the JSFifddle breaks the `$translateProvider` functions (I got it somehow fixed by extending the replacing variable like so:` $TranslateProvider = angular.extend(m, $TranslateProvider);` after injecting` $translateProvider` as `m` and add `pascalprecht.translate` as a module dependency, and add `var` declarations to be able to `'use strict'`) - so right off the bat it doesn't seem to be a solid solution.

With that JSFiddle code, I got the _directive_ to work, at least for non-nested cases (which are fine and all, but not what's needed here, so I didn't bother testing the nested cases) but not the filter (which I need working for both nested and non-nested translations).

It seems to me namespaced translations is a pretty big deal, and therefore should be available for all the 3 translation methods (service, directive and filter) - and that I'm doing something wrong somewhere.

"service" method is rather limited ($translate service doesn't provide a two-way data binding which means more code to listen to events as per the documentation and, well, the fact that loading all the interface's string translation in the app's controller/service seems to be a pretty bad practice ; views/templates are here for that if I'm not mistaken).

Is there something I'm not doing or understanding quite well, or is that angular-translate nested JSON works only using the `$translate` service?

**Comment by froger-me** [comment]
That seems to do the trick - although I'm really not sure about performance with that cache thing.
Inspired from http://embed.plnkr.co/EK2TYI1NZevojOFDpaOG and https://glebbahmutov.com/blog/async-angular-filter/

Maybe there would be a way to clear that cache variable once it grows big?

```
// Including ngTranslate
angular.module("ngTranslate", ["ng", "pascalprecht.translate"])
.config(["$provide", "$translateProvider", function (n, m) {
    var $TranslateProvider = function () {
        var n, t = {};
        this.translations = function (n, r) {
            if (!n && !r) return t;
            if (n && !r) {
                if (angular.isString(n)) return t[n];
                t = n
            } else t[n] = r
        }, this.uses = function (r) {
            if (!r) return n;
            if (!t[r]) throw Error("$translateProvider couldn't find translationTable for langKey: '" + r + "'");
            n = r
        }, this.$get = ["$interpolate", "$log", function (r, a) {
            return $translate = function (e, i) {
                var l = n ? t[n][e] : t[e];
                return l ? r(l)(i) : (a.warn("Translation for " + e + " doesn't exist"), e)
            }
        }]
    };
    $TranslateProvider = angular.extend(m, $TranslateProvider);
    n.provider("$translate", $TranslateProvider)
}]);

angular.module("ngTranslate").directive("translate", ["$filter", "$interpolate", function (n, t) {
    var r = n("translate");
    return {
        restrict: "A",
        scope: !0,
        link: function (n, a, e) {
            e.$observe("translate", function (r) {
                n.translationId = angular.equals(r, "") ? t(a.text())(n.$parent) : r
            }), e.$observe("values", function (t) {
                n.interpolateParams = t
            }), n.$watch("translationId + interpolateParams", function () {
                a.html(r(n.translationId, n.interpolateParams))
            })
        }
    }
}]);

angular.module("ngTranslate").filter("translate", ["$parse", "$translate", function ($parse, $translate) {

	var cached = {};
    function translateFilter (translationId, interpolateParams) {

    	if (translationId) {
    		if (translationId in cached) {
    			return typeof cached[translationId].then !== 'function' ? cached[translationId] : undefined;
    		} else {
    			if (!angular.isObject(interpolateParams)) {
    				interpolateParams = $parse(interpolateParams)();
    			}
				$translate(translationId, interpolateParams).then( function (translation) {
					cached[translationId] = translation;
				});
    		}
    	}
    }
	translateFilter.$stateful = true;
    return translateFilter;
}]);
```

**Comment by froger-me** [comment]
does not work with messageFormat interpolation: using a SAP, change state, come back to previous state, interpolation in message format isn't done and it displays the uninterpreted expression - Darn!

**Comment by froger-me** [comment]
That seems enough - I'd like to point out I'm new to angular.
```
angular.module("pascalprecht.translate").filter("translate", ["$parse", "$translate", function ($parse, $translate) {

    var cached = {};
    function translateFilter (translationId, interpolateParams) {

        if (translationId) {
            // seems very hack-ish
            if (translationId in cached) {
                // seems very hack-ish
                var result = (typeof cached[translationId].then !== 'function') ? cached[translationId] : undefined;

                // seems very hack-ish
                if (Object.keys(cached).length > 100) {
                    cached = {};
                }
                return result;
            } else {
                if (!angular.isObject(interpolateParams)) {
                    interpolateParams = $parse(interpolateParams)();
                }
                // in truth, the whole thing seems hack-ish
                $translate(translationId, interpolateParams).then( function (translation) {
                    cached[translationId] = translation;
                });
            }
        }
    }
    translateFilter.$stateful = true;
    return translateFilter;
}]);
```

**Comment by knalli** [comment]
Side note: AngularJS does caching filter results itself, why should you implement this on your own?

Also, this has nothing to do with you original problem with the filter?

**Comment by froger-me** [comment]
Well, the fact is, as I said, I have very little idea where I'm going with that (=> "I'd like to point out I'm new to angular.").
After trying for hours to make the filter work with nested translations, I thought I'd open an issue to ask for advice, and meanwhile that wouldn't prevent me from trying and solve the issue I have on my own. Obviously I am prepared to receive feedback pointing out I said a lot of irrelevant things and went in bad directions -_-'.

Now, it seems I did something wrong form the beginning, because the nested translations work without any changes. I do not know what I did wrong, but I'm sure it was because all that code is useless, the translations are displayed properly now. It's not an issue to make mistakes, but not knowing where is annoying.

Long story short, seems it's a false alarm.
